### PR TITLE
Support adding directories in google cloud storage remote

### LINKED ISSUE #2814: gs: support directories as external dependencies/outputs
I have a similar issue to #2678 but for GS.

I have a bucket with the following structure
```
my_bucket
       ├── data
       │     ├── img1.png
       │     ├── img2.png
       │     ├── ...
       └── cache
```

I have then created a clean project

```
$ git init
$ dvc init
$ dvc remote add gscache gs://my_bucket/cache
$ dvc config cache.gs gscache
$ dvc add gs://my_bucket/data
```

The output is as follows:
```
100%|██████████|Add                                                                                                                            1/1 [00:00<00:00,  1.21file/s]
ERROR: output 'gs://my_bucket/data' does not exist
```

Adding a single file works (`dvc add gs://my_bucket/data/img1.png`).

A more verbose version:
```
$ dvc add gs://my_bucket/data -v 
DEBUG: PRAGMA user_version;
DEBUG: fetched: [(3,)]
DEBUG: CREATE TABLE IF NOT EXISTS state (inode INTEGER PRIMARY KEY, mtime TEXT NOT NULL, size TEXT NOT NULL, md5 TEXT NOT NULL, timestamp TEXT NOT NULL)
DEBUG: CREATE TABLE IF NOT EXISTS state_info (count INTEGER)
DEBUG: CREATE TABLE IF NOT EXISTS link_state (path TEXT PRIMARY KEY, inode INTEGER NOT NULL, mtime TEXT NOT NULL)
DEBUG: INSERT OR IGNORE INTO state_info (count) SELECT 0 WHERE NOT EXISTS (SELECT * FROM state_info)
DEBUG: PRAGMA user_version = 3;
100%|██████████|Add                                                                                                                            1/1 [00:01<00:00,  1.63s/file]
DEBUG: SELECT count from state_info WHERE rowid=?
DEBUG: fetched: [(0,)]
DEBUG: UPDATE state_info SET count = ? WHERE rowid = ?
ERROR: output 'gs://my_bucket/data' does not exist
------------------------------------------------------------
Traceback (most recent call last):
  File "/home/egnyte/anaconda3/envs/dvc/lib/python3.7/site-packages/dvc/command/add.py", line 25, in run
    fname=self.args.file,
  File "/home/egnyte/anaconda3/envs/dvc/lib/python3.7/site-packages/dvc/repo/__init__.py", line 35, in wrapper
    ret = f(repo, *args, **kwargs)
  File "/home/egnyte/anaconda3/envs/dvc/lib/python3.7/site-packages/dvc/repo/scm_context.py", line 4, in run
    result = method(repo, *args, **kw)
  File "/home/egnyte/anaconda3/envs/dvc/lib/python3.7/site-packages/dvc/repo/add.py", line 53, in add
    stage.save()
  File "/home/egnyte/anaconda3/envs/dvc/lib/python3.7/site-packages/dvc/stage.py", line 716, in save
    out.save()
  File "/home/egnyte/anaconda3/envs/dvc/lib/python3.7/site-packages/dvc/output/base.py", line 219, in save
    raise self.DoesNotExistError(self)
dvc.output.base.OutputDoesNotExistError: output 'gs://my_bucket/data' does not exist
------------------------------------------------------------
```

`dvc --version = 0.68.1`. I am using ubuntu, I installed using conda, python 3.7.5.

### PR DESCRIPTION
**Pull request by skshetry**
* [x] ❗ Have you followed the guidelines in the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) list?

* [x] 📖 Check this box if this PR **does not** require [documentation](https://dvc.org/doc) updates, or if it does **and** you have created a separate PR in [dvc.org](https://github.com/iterative/dvc.org) with such updates (or at least opened an issue about it in that repo). Please link below to your PR (or issue) in the [dvc.org](https://github.com/iterative/dvc.org) repo.

* [x] ❌ Have you checked DeepSource, CodeClimate, and other sanity checks below? We consider their findings recommendatory and don't expect everything to be addressed. Please review them carefully and fix those that actually improve code or fix bugs.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

#### Description
I have reused test code from `unit/test_s3.py` by moving it to `functional` tests, as, I argue, moto server can be assumed as functional as real `s3` (which is a selling point of `moto` itself). The same tests are run with different remotes (i.e. `s3` and `gs`).

And, I have *stolen* most of the code from #2619.

The logic behind `s3` and `gs` is almost similar, as they both are object storages. So, there is an opportunity for abstraction for `ObjectStorageRemote`, but, I think that it'll be over-engineering for now (what about `azure`?).

#### Todo
- [ ] Reuse code instead of copy-pasting. 
- [x] Check for `empty_dir`. Will that work?
- [x] Use prefix for the test, or custom bucket (might not be possible locally)
- [x] Check if tests for s3 and gs can be combined with different remotes.
- [x] Rebase with current master.
- [ ] Run some manual tests with directories.
- [x] Run all tests locally 2-3 times with `DVC_TEST_GCP` set and make it pass.
- [x] Make this PR ready for review.

#### How was this tested?
1. Upload folder in structure similar to the following:
```
$ gsutil ls -r gs://dvc-test/skshetry-test
gs://dvc-test/skshetry-test/:
gs://dvc-test/skshetry-test/data1.txt

gs://dvc-test/skshetry-test/data/:
gs://dvc-test/skshetry-test/data/data2.txt
gs://dvc-test/skshetry-test/data/data3.txt
gs://dvc-test/skshetry-test/data/data5.txt

gs://dvc-test/skshetry-test/data/subdir/:
gs://dvc-test/skshetry-test/data/subdir/data4.txt

gs://dvc-test/skshetry-test/data/subdir2/:
gs://dvc-test/skshetry-test/data/subdir2/data6.txt
gs://dvc-test/skshetry-test/data/subdir2/data7.txt

gs://dvc-test/skshetry-test/data/subdir3/:

gs://dvc-test/skshetry-test/data/subdir3/nested/:
gs://dvc-test/skshetry-test/data/subdir3/nested/data6.txt
```

2.
```
$ git init
$ dvc init
$ dvc remote add gscache gs://dvc-test/skshetry-test
$ dvc config cache.gs gscache
```

3. `dvc add gs://dvc-test/skshetry-test/data` (check `gs://dvc-test/skshetry-test/data1.txt` should not get uploaded).
4. Reset cache, and `dvc add gs://dvc-test/skshetry-test/data/subdir`.
5. Similar to step 4 with `dvc add gs://dvc-test/skshetry-test/data/subdir2`.
6. Similar to step 4 with `dvc add gs://dvc-test/skshetry-test/data/subdir3`.
7. Similar to step 4 with `dvc add gs://dvc-test/skshetry-test` and `dvc add gs://dvc-test/skshetry-test/`.
8. Reset cache and `dvc run -d remote://gs/data 'echo hello world'`.

Fixes #2814

### CODE REVIEW

File: tests/func/test_gs.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        remote.gs, "dvc-test", str(random_path_prefix) + "/data/subdir/2", "2"
+    )
+    create_object(
+        remote.gs, "dvc-test", str(random_path_prefix) + "/data/subdir/3", "3"
+    )
+
+    yield remote
+
+    for file in remote.walk_files(remote.path_info / str(random_path_prefix)):
+        remote.remove(file)
```
Comment (skshetry) [inline tests/func/test_gs.py]: - [ ] Remove the use of `walk_files` from here.

**Comment by Suor** [comment]
Am I missing something? We agreed to not add this for now. The ticket https://github.com/iterative/dvc/issues/1654 is closed.

**Comment by skshetry** [comment]
@suor, there is another issue for `gs` only: #2814. Forgot to mention in the PR.

**Comment by efiop** [comment]
FYI: added [WIP] prefix to make it more obvious ("draft" is fine, but prefix seems to be a bit easier to grasp, no biggie) 🙂

File: dvc/remote/gs.py [hunk 2]
```diff
@@ -136,16 +136,33 @@ def remove(self, path_info):
...
-        for blob in self.gs.bucket(bucket).list_blobs(prefix=prefix):
+    def _list_paths(self, bucket, prefix, max_items=None):
+        for blob in self.gs.bucket(bucket).list_blobs(
+            prefix=prefix, max_results=max_items
+        ):
             yield blob.name
 
     def list_cache_paths(self):
-        return self._list_paths(self.path_info.bucket, self.path_info.path)
+        return self.walk_files(self.path_info)
```
Comment (skshetry) [inline dvc/remote/gs.py]: Ooops, tests are failing due to this for `gs`. `walk_files` provides absolute urlpath, whereas `_list_paths` provided relative path.

Comment (skshetry) [reply dvc/remote/gs.py]: Fixed by 54dd9890dc83b

Comment (skshetry) [reply dvc/remote/gs.py]: It's failing again, needs more investigation.

Comment (skshetry) [reply dvc/remote/gs.py]: `tests/unit/test_ignore.py` seems to be always failing after running tests once. I have run all the tests 10 times, and it's always the 6 scenarios in that file that are failing. Other tests (with `DVC_TEST_GCP` are passing).

Comment (efiop) [reply dvc/remote/gs.py]: @skshetry I'm not able to reproduce `test_ignore.py` failing on master or on your branch 🤔

File: tests/func/test_remote_dir.py [hunk 3]
```diff
@@ -1,12 +1,25 @@
 # -*- coding: utf-8 -*-
```
Comment (Suor) [inline tests/func/test_remote_dir.py]: This is still units tests since we don't make a repo and don't start all the dvc machinery.

**Comment by efiop** [comment]
@skshetry About empty dirs, have you seen https://github.com/iterative/dvc/issues/2678 ? Does gs have a concept of empty dirs? 🙂

File: tests/func/test_remote_dir.py [hunk 4]
```diff
@@ -38,7 +51,54 @@ def remote():
...
+
+    yield remote
+
+    for teardown in teardowns:
+        teardown()
+
+
+remote_parameterized = pytest.mark.parametrize(
+    "remote_name", [pytest.param("remote_gs", marks=test_gs), "remote_s3"]
+)
```
Comment (Suor) [inline tests/func/test_remote_dir.py]: We are starting to duplicate whatever we have in `tests/func/test_api.py`. Let's extract remote definitions from there to `tests/remotes.py` and supplement that with whatever you need here. E.g. with `.get_remote()` and/or `.put_object()`.

Then make a `remote` fixture:
```python
@pytest.fixture
def remote(request):
    if not request.param.should_test():
        raise pytest.skip()
    return request.param.get_remote()
```

File: tests/func/test_remote_dir.py [hunk 5]
```diff
@@ -38,7 +51,54 @@ def remote():
...
+        teardowns.append(lambda: remote.remove(remote.path_info / file))
+
+    put_object("empty_dir/", "")
+    put_object("empty_file", "")
+    put_object("foo", "foo")
+    put_object("data/alice", "alice")
+    put_object("data/alpha", "alpha")
+    put_object("data/subdir/1", "1")
+    put_object("data/subdir/2", "2")
+    put_object("data/subdir/3", "3")
```
Comment (Suor) [inline tests/func/test_remote_dir.py]: This duplicates the structure. And since the tests are the same these two need to stay in sync, let's refactor it somehow to remote duplication.

Maybe the following - remote desc classes from `test_api.py` - will help.

**Comment by skshetry** [comment]
> @skshetry About empty dirs, have you seen #2678 ? Does gs have a concept of empty dirs?

Yep, I saw that, and this also suffers from same issue.
```
➜ dvc add gs://dvc-test/skshetry-test/empty-dir/
100%|██████████|Add                                                                                                 1.00/1.00 [00:21<00:00,  21.1s/file]
ERROR: 'skshetry-test/empty-dir/.' doesn't exist in the cloud 
```

`gsutil` does not support creating empty directories, but [Google Cloud Console seems to support this](https://cloud.google.com/storage/docs/quickstart-console#create_folders) which I can't confirm as I don't have access to it. And, we can `emulate` this by creating blob with trailing slash ([`gsutil` provides an illusion to this](https://cloud.google.com/storage/docs/gsutil/addlhelp/HowSubdirectoriesWork).)

I didn't act on it because of https://github.com/iterative/dvc/pull/2683#issuecomment-548125068, as I am not really sure what you mean by solving it properly. I guess, I was just waiting for how it will be solved for `s3`.

File: dvc/remote/base.py [hunk 6]
```diff
@@ -312,7 +312,7 @@ def get_checksum(self, path_info):
             return checksum
 
         if self.isdir(path_info):
-            checksum = self.get_dir_checksum(path_info)
+            checksum = self.get_dir_checksum(path_info / "")
```
[the remark this answers is not in the document]
Comment (skshetry) [reply dvc/remote/base.py]: This does not leak any GS internals. This could have been solved by adding it to methods inside `RemoteGS`, but, I made a compromise between handling it higher vs looking into each remotes and fixing there.

> Why do we need this?
Say, you have `gs://dvc-test/data1.txt` file and `gs://dvc-test/data/data2.txt`. If I do `dvc add gs://dvc-test/data`, it will try to `walk_file` which will include `gs://dvc-test/data1.txt` as well.

> Another questions is why don't we need this in S3? Is there a bug?
Not sure, if this is an issue in real, quick addition of `s3.put_object(Bucket="bucket", Key="data1.txt", Body=b"3")` in tests did fail the `test_walk_files` for s3. Needs further investigation. Probably a bug.

File: dvc/remote/gs.py [hunk 7]
```diff
@@ -136,16 +136,34 @@ def remove(self, path_info):
...
+    def _list_paths(self, bucket, prefix, max_items=None):
+        for blob in self.gs.bucket(bucket).list_blobs(
+            prefix=prefix, max_results=max_items
+        ):
             yield blob.name
 
     def list_cache_paths(self):
-        return self._list_paths(self.path_info.bucket, self.path_info.path)
+        for cache in self.walk_files(self.path_info):
+            yield cache.path
```
[the remark this answers is not in the document]
Comment (skshetry) [reply dvc/remote/gs.py]: I feel, `walk_files` is a *public* api. So, just trying to not to use *private* api. What do you say?

File: dvc/remote/gs.py [hunk 8]
```diff
@@ -136,16 +136,34 @@ def remove(self, path_info):
 
         blob.delete()
 
-    def _list_paths(self, bucket, prefix):
-        for blob in self.gs.bucket(bucket).list_blobs(prefix=prefix):
+    def _list_paths(self, bucket, prefix, max_items=None):
```
[the remark this answers is not in the document]
Comment (skshetry) [reply dvc/remote/gs.py]: Thanks, will do.

File: tests/func/test_remote_dir.py (hunk 4 above)
Comment (skshetry) [reply tests/func/test_remote_dir.py]: Ah, did not know that. Will do.

File: tests/func/test_remote_dir.py (hunk 3 above)
Comment (skshetry) [reply tests/func/test_remote_dir.py]: Agreed.

**Comment by efiop** [comment]
For the record, had an additional discussion on discord: https://discordapp.com/channels/485586884165107732/565699007037571084/649634041993101333

File: tests/func/test_remote_dir.py [hunk 9]
```diff
@@ -38,7 +51,55 @@ def remote():
...
+    put_object("data/subdir/4/", "")
```
Comment (skshetry) [inline tests/func/test_remote_dir.py]: I added this one, so that we could test nested directories, and demonstrate that it is not `walked`.

File: dvc/remote/gs.py [hunk 10]
```diff
@@ -136,16 +136,42 @@ def remove(self, path_info):
...
+        for fname in self._list_paths(path_info.bucket, path_info.path):
+            # skip nested empty directories
+            if fname.endswith("/"):
+                continue
+            yield path_info.replace(fname)
+
+    def makedirs(self, path_info):
+        self.gs.bucket(path_info.bucket).blob(
+            (path_info / "").path
+        ).upload_from_string("")
```
Comment (skshetry) [inline dvc/remote/gs.py]: I might be missing some bigger picture here, but this will create a directory if it doesnot exist when checking out.

File: tests/func/test_remote_dir.py [hunk 11]
```diff
@@ -38,7 +51,54 @@ def remote():
...
+        ├── empty_dir
+        ├── empty_file
+        └── foo
+    """
+    prefix = str(uuid.uuid4())
+    REMOTE_URL = "gs://dvc-test/" + prefix
+    remote = RemoteGS(None, {"url": REMOTE_URL})
+    teardowns = []
+
+    def put_object(file, content):
```
[the remark this answers is not in the document]
Comment (skshetry) [reply tests/func/test_remote_dir.py]: I keep forgetting about `2.7`. :grin:

File: dvc/remote/gs.py (hunk 10 above)
Comment (efiop) [reply dvc/remote/gs.py]: @skshetry Very good point! So we won't "create" empty dirs inside of the added directory, as there won't be particular files to checkout in there. And the top level empty dir will be created anyway. That sounds like a great way to handle this without additional fuss.  CC @mroutis

File: dvc/remote/gs.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (efiop) [reply dvc/remote/gs.py]: @skshetry Not sure I get your point. What is wrong with using a private method inside of a public one?

File: dvc/remote/base.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (efiop) [reply dvc/remote/base.py]: @skshetry I agree with @Suor , this change looks very weird. `get_dir_checksum` should simply accept `path_info`, you area leaking "empty_dir" concept here. Effectively you are feeding it an empty_dir object and not the directory as a whole, which might be confusing and error prone. s3 didn't have to do this, so I'm pretty sure you don't have to do this as well.

[the remark this answers is not in the document]
Comment (mroutis) [reply dvc/remote/base.py]: > Say, you have gs://dvc-test/data1.txt file and gs://dvc-test/data/data2.txt. If I do dvc add gs://dvc-test/data, it will try to walk_file which will include gs://dvc-test/data1.txt as well.

@skshetry, yep, this is due to S3 and GCS searching by prefix instead of paths, I think it is remote's problem.

File: dvc/remote/gs.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mroutis) [reply dvc/remote/gs.py]: @skshetry , private methods can be called inside the module/class they have been defined

File: dvc/remote/gs.py (hunk 10 above)
Comment (mroutis) [reply dvc/remote/gs.py]: @efiop, I'm not sure if we should do that :thinking: 

I'd need to take a look, but so far, it looks like such compromise is not that bad, so yeah, maybe adding a `makedirs` should work.

File: tests/func/test_remote_dir.py (hunk 9 above)
Comment (mroutis) [reply tests/func/test_remote_dir.py]: @skshetry , what about `"data/subdir/empty_dir/"`?

File: dvc/remote/gs.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (skshetry) [reply dvc/remote/gs.py]: @efiop, @mroutis, it's more to do with my comfort working with *public* api (they might be more *tested*/*used*). I don't have a strong opinion here, and y'all point is more valid.

File: tests/func/test_remote_dir.py (hunk 9 above)
Comment (skshetry) [reply tests/func/test_remote_dir.py]: Sorry, I do not understand. I understood it as to, why am I creating a file with content `""` (empty quotes), right? Refer here: https://stackoverflow.com/a/38417397

If not, please clarify.

Comment (mroutis) [reply tests/func/test_remote_dir.py]: @skshetry, no, it was more like giving a more meaningful name to an empty file,

I suggested changing `4/` to `empty_dir/`, if it is only a file (without the trailing slash), you can try `empty_file` or just `empty`.

File: dvc/remote/gs.py (hunk 10 above)
Comment (mroutis) [reply dvc/remote/gs.py]: https://github.com/iterative/dvc/pull/2683

@skshetry, indeed, adding `makedirs` fixed the problem right away, wonderful catch!

File: dvc/remote/gs.py [hunk 12]
```diff
@@ -138,16 +138,35 @@ def remove(self, path_info):
...
+    def isdir(self, path_info):
+        dir_path = path_info / ""
+        return bool(list(self._list_paths(dir_path, max_items=1)))
 
     def exists(self, path_info):
-        paths = set(self._list_paths(path_info.bucket, path_info.path))
-        return any(path_info.path == path for path in paths)
+        dir_path = path_info / ""
+        fname = next(self._list_paths(path_info, max_items=1), "")
+        return path_info.path == fname or fname.startswith(dir_path.path)
```
Comment (skshetry) [inline dvc/remote/gs.py]: @efiop, this two functions are identical with that in s3, and therefore, codeclimate is failing. Can you suggest me anything? Or, ignore for now?

File: tests/func/test_remote_dir.py (hunk 4 above)
Comment (skshetry) [reply tests/func/test_remote_dir.py]: @Suor, I have refactored these, but, I went with using `contextmanager` so not to create any track any *states*. Please review again.

File: dvc/remote/gs.py (hunk 12 above)
Comment (efiop) [reply dvc/remote/gs.py]: @skshetry I would just ignore it for now. 🙂

Comment (skshetry) [inline dvc/remote/gs.py]: Just found out that this logic is wrong.

Say, we have a file `data/subdir-file.txt` and `data/subdir/1`. When we go add `data/subdir`, it will find substr for `subdir-file.txt`, but won't really find `subdir` as there's no check for `isdir`. And, therefore, it would fail with a message that the `'data/subdir' does not exist`.

Comment (skshetry) [reply dvc/remote/gs.py]: Same logic is in `S3`

Comment (Suor) [reply dvc/remote/gs.py]: This line `dir_path = path_info / ""` adds a trailing `/` to `dir_path`, so `subdir-file.txt` won't start with `dir_path.path`.

File: tests/remotes.py [hunk 13]
```diff
@@ -0,0 +1,337 @@
+import os
```
Comment (Suor) [inline tests/remotes.py]: Several high level considerations here:

1. This is cool, this was waiting to be factored out for quite some time.
2. Are you sure you want to do that in this PR? You may come across some issues while unifying this that you don't expect now. 
3. Things like `_should_test_aws()` and `get_ssh_url()` should be merged into classes, e.g. move to `S3.should_test()` and `SSH.get_url()`.

File: tests/remotes.py [hunk 14]
```diff
@@ -0,0 +1,337 @@
...
+
+    @classmethod
+    def remote(cls):
+        @contextmanager
+        def inner():
+            with mock_s3():
+                remote = RemoteS3(None, {"url": cls.get_url()})
+                yield remote
+
+        return inner()
```
Comment (Suor) [inline tests/remotes.py]: I don't think this nested is needed.

File: dvc/remote/gs.py (hunk 12 above)
Comment (skshetry) [reply dvc/remote/gs.py]: Yes, sorry, I did not explain it correctly. The `next()` could provide you `subdir-file.txt` which won't match, therefore, it returning False, resulting in `'data/subdir' does not exist` message.

There's no short-circuiting for checking if a given file/folder exists. You have to first check if the folder exists, and then, loop through it until you find exact match.

**Comment by Suor** [comment]
Wet should list all, not only first one indeed. Good catch.

сб, 30 нояб. 2019 г., 13:43 saugat pachhai <notifications@github.com>:

File: tests/remotes.py (hunk 13 above)
Comment (skshetry) [reply tests/remotes.py]: > Are you sure you want to do that in this PR? You may come across some issues while unifying this that you don't expect now.

This PR has gotten quite large. So, I'd say, I can work on *nitpicky* changes on a separate PR. 
If it's a blocker, I'd definitely fix those, though. The tests have all passed. So, I guess, there won't be further issues.

> Things like _should_test_aws() and get_ssh_url() should be merged into classes, e.g. move to S3.should_test() and SSH.get_url().
Can I do that refactor in a separate PR? I want just wanted to move things around for this PR, so, I'd opt not to merge those in this PR itself.

P.S. I haven't also added tests as it will collide with #2873, which, I'll do in a separate PR.

Comment (Suor) [reply tests/remotes.py]: You may postpone some things, just make an issue for it and refer to a comment/discussion here.

Comment (skshetry) [reply tests/remotes.py]: Can you review again?  :)

**Comment by efiop** [comment]
For the record: windows tests are failing due to unrelated problems with chocolatey packages. The rest of the tests are fine. ;)

**Comment by Suor** [comment]
Wrote about listing for exists here - https://github.com/iterative/dvc/pull/2873#discussion_r352299565. TLDR, the reason why we do listing instead of some `Blob.exists()` call is unknown. If we may do the latter then we won't have all those issues.

BTW, we will still need to list to check if dir exists, which is not good for cache batch exists, quoting from there:
> Additional consideration about cache, we don't really want to check `.isdir()` for every checksum, but we want `.exists()` to work for both files and dirs for remotes supporting that. This contradiction might be resolved by calling `.isfile()` instead of `.exists()` inside `.cache_exists()`. We will need to make `isfile()` work properly though, now it is just a stab in `RemoteBASE`.

**Comment by skshetry** [comment]
Thanks @Suor. I was too pinned on checking for the directory, that I missed it. I implemented `isfile`, but for the `.cache_exists()`, we could implement that in #2873 or in a separate PR.

File: tests/remotes.py [hunk 15]
```diff
@@ -0,0 +1,325 @@
...
+        bucket = client.get_bucket(bucket)
+        blob = bucket.blob(remote.path_info.path + "/" + key)
+        blob.upload_from_string(body)
+
+    @staticmethod
+    def get_client(remote):
+        return remote.gs
+
+    @staticmethod
+    def remove(remote, files):
```
Comment (Suor) [inline tests/remotes.py]: We use random prefixes on a bucket, which is cleared periodically, so we shouldn't waste time removing blobs here.

File: tests/remotes.py [hunk 16]
```diff
@@ -0,0 +1,325 @@
...
+    @contextmanager
+    def put_objects(cls, remote, objects):
+        s3 = cls.get_client(remote)
+        s3.create_bucket(Bucket="dvc-test")
+        for key, body in objects.items():
+            cls.put_object(remote, key, body)
+        yield
+
+    @staticmethod
+    def get_client(remote):
```
Comment (Suor) [inline tests/remotes.py]: I would just inline it. All methods calling it is remote specific anyway.

File: tests/remotes.py [hunk 17]
```diff
@@ -0,0 +1,325 @@
...
+        return remote.s3
+
+    @classmethod
+    def put_object(cls, remote, key, body):
+        s3 = cls.get_client(remote)
+        bucket = remote.path_info.bucket
+
+        s3.put_object(
+            Bucket=bucket, Key=remote.path_info.path + "/" + key, Body=body
```
Comment (Suor) [inline tests/remotes.py]: To correctly join paths we use `(remote.path_info / key).path`. This was one of the reasons `PathsInfo/URLInfo` were introduced some time ago)

File: tests/remotes.py [hunk 18]
```diff
@@ -0,0 +1,325 @@
...
+        for key, body in objects.items():
+            cls.put_object(remote, key, body)
+        yield
+
+    @staticmethod
+    def get_client(remote):
+        return remote.s3
+
+    @classmethod
+    def put_object(cls, remote, key, body):
```
Comment (Suor) [inline tests/remotes.py]: Do we need a separate method for that? Consider inlining it.

File: tests/remotes.py [hunk 19]
```diff
@@ -0,0 +1,325 @@
...
+    @classmethod
+    @contextmanager
+    def remote(cls):
+        with mock_s3():
+            remote = RemoteS3(None, {"url": cls.get_url()})
+            yield remote
+
+    @classmethod
+    @contextmanager
+    def put_objects(cls, remote, objects):
```
Comment (Suor) [inline tests/remotes.py]: Why do we make it a context manager? a simple function should do the job.

File: tests/remotes.py [hunk 20]
```diff
@@ -0,0 +1,325 @@
...
+        for key, body in objects.items():
+            cls.put_object(remote, key, body)
+        yield
+        cls.remove(remote, objects.keys())
+
+    @classmethod
+    def put_object(cls, remote, key, body):
```
Comment (Suor) [inline tests/remotes.py]: Also consider inlining it. We redo things like `get_client()` and `get_bucket()` here. The real for loop body is just:
```python
bucket.blob((remote.path_info / key).path).upload_from_string(body)
```

File: tests/remotes.py [hunk 21]
```diff
@@ -0,0 +1,325 @@
...
+    def remote(cls):
+        with mock_s3():
+            remote = RemoteS3(None, {"url": cls.get_url()})
+            yield remote
+
+    @classmethod
+    @contextmanager
+    def put_objects(cls, remote, objects):
+        s3 = cls.get_client(remote)
+        s3.create_bucket(Bucket="dvc-test")
```
Comment (Suor) [inline tests/remotes.py]: How does this bucket corresponds to `.get_url()`? Maybe you need to change `.get_url()` to return a const?

File: tests/unit/remote/test_remote_dir.py [hunk 22]
```diff
@@ -72,20 +66,21 @@ def test_exists(remote):
...
-        remote.path_info / "foo",
+        remote.path_info / "data/subdir/empty_file",
     ]
 
-    assert list(remote.walk_files(remote.path_info)) == files
+    assert list(remote.walk_files(remote.path_info / "data")) == files
 
 
+@pytest.mark.parametrize("remote", [S3Mocked], indirect=True)
 def test_copy_preserve_etag_across_buckets(remote):
```
Comment (Suor) [inline tests/unit/remote/test_remote_dir.py]: We don't need to create all those blobs for this test and `parametrize` looks a bit stretched. I suggest using `@mock_s3` and creating both remotes simply with `RemoteS3(...)` and not using `S3Mocked` here.

File: tests/remotes.py (hunk 15 above)
Comment (skshetry) [reply tests/remotes.py]: ah, didn't know that. Okay, then, there does not need to be any context managers for `put_objects`.

File: tests/unit/remote/test_remote_dir.py (hunk 22 above)
Comment (Suor) [reply tests/unit/remote/test_remote_dir.py]: Current approach is also fine, I guess.

File: tests/remotes.py (hunk 21 above)
Comment (Suor) [reply tests/remotes.py]: And anyway you can use `remote.path_info.bucket` to not repeat yourself.

File: tests/unit/remote/test_remote_dir.py (hunk 22 above)
Comment (skshetry) [reply tests/unit/remote/test_remote_dir.py]: Keeping it this way for now, then. :)

File: tests/remotes.py (hunk 21 above)
Comment (skshetry) [reply tests/remotes.py]: yes, did the same already. :)

File: tests/unit/remote/test_remote_dir.py [hunk 23]
```diff
@@ -97,15 +92,39 @@ def test_copy_preserve_etag_across_buckets(remote):
...
+        (True, "data/subdir/3"),
+        (False, "data/subdir/empty_dir/"),
+        (True, "data/subdir/empty_file"),
+        (False, "something-that-does-not-exist"),
+        (False, "data/subdir/empty-file/"),
+        (False, "empty_dir"),
+    ]
+
+    for expected, path in test_cases:
+        assert remote.isfile(remote.path_info / path) == expected
```
Comment (skshetry) [inline tests/unit/remote/test_remote_dir.py]: This test will need adjustment if #2873 gets merged.

Comment (efiop) [reply tests/unit/remote/test_remote_dir.py]: Yeah, let's merge #2873 first, once we get one more approval 😉 And then we'll merge this one.

File: dvc/remote/gs.py [hunk 24]
```diff
@@ -138,16 +138,45 @@ def remove(self, path_info):
...
 
     def exists(self, path_info):
-        paths = set(self._list_paths(path_info.bucket, path_info.path))
-        return any(path_info.path == path for path in paths)
+        """Check if the blob exists. If it does not exist,
+        it could be a part of a directory path.
+
+        eg: if `data/file.txt` exists, check for `data` should return True
+        """
+        return self.isfile(path_info) or self.isdir(path_info / "")
```
Comment (Suor) [inline dvc/remote/gs.py]:
```suggestion
        return self.isfile(path_info) or self.isdir(path_info)
```
No need to make it twice.

File: tests/unit/remote/test_remote_dir.py (hunk 23 above)
Comment (efiop) [reply tests/unit/remote/test_remote_dir.py]: @skshetry  #2873 is merged, please adjust this PR accordingly 🙂

**Comment by Suor** [comment]
BTW, need to create `.cache_exists()/isfile()` issue before closing it.

**Comment by skshetry** [comment]
@Suor, I have created an issue: #2877. Regarding the suggestion https://github.com/iterative/dvc/pull/2853#discussion_r352283370, I'll make a issue after this gets merged.

**Comment by efiop** [comment]
Great stuff, thank you so much @skshetry ! 🚀 

For the record: windows tests are failing due to temporary problems with chocolatey. Ignoring that for now.

File: tests/remotes.py (hunk 13 above)
Comment (skshetry) [reply tests/remotes.py]: > Things like _should_test_aws() and get_ssh_url() should be merged into classes, e.g. move to S3.should_test() and SSH.get_url().

Created #2878 to work on this further.
